presto-main/src/test/java/com/facebook/presto/operator/TestGroupByHash.java

```diff
@@ -293,4 +299,90 @@ else if (type == BIGINT) {
         long[] expected = IntStream.range(0, actual.length).mapToLong(i -> entryBytes * expectedSize << (i + 1)).toArray();
         assertEquals(actual, expected);
     }
+
+    @Test(dataProvider = "dataType")
+    public void testPageYield(Type type)
+            throws Exception
+    {
+        // Create a page with positionCount >> expected size of groupByHash
+        int length = 1_000_000;
+        Block valuesBlock;
+        if (type == VARCHAR) {
+            valuesBlock = createStringSequenceBlock(0, length);
+        }
+        else if (type == BIGINT) {
+            valuesBlock = createLongSequenceBlock(0, length);
+        }
+        else {
+            throw new IllegalArgumentException("unsupported data type");
+        }
+        Block hashBlock = getHashBlock(ImmutableList.of(type), valuesBlock);
+        Page page = new Page(valuesBlock, hashBlock);
+        AtomicInteger counter = new AtomicInteger();
+        ExtraMemoryReservation forceYield = extraMemoryInBytes -> counter.incrementAndGet() % 3 == 0;
+        int yields = 0;
+
+        // test addPage
+        GroupByHash groupByHash = createGroupByHash(TEST_SESSION, ImmutableList.of(type), new int[] {0}, Optional.of(1), 4, JOIN_COMPILER, forceYield);
+
+        Work<?> addPageWork = groupByHash.addPage(page);
+        while (!addPageWork.process()) {
+            yields++;
+        }
+        assertEquals(length, groupByHash.getGroupCount());
+        assertGreaterThanOrEqual(yields, (int) log(length), "at least log(length) many rehash");
+
+        // test getGroupIds
+        counter.set(0);
+        yields = 0;
+        groupByHash = createGroupByHash(TEST_SESSION, ImmutableList.of(type), new int[] {0}, Optional.of(1), 4, JOIN_COMPILER, forceYield);
+
+        Work<GroupByIdBlock> getGroupIdsWork = groupByHash.getGroupIds(page);
+        while (!getGroupIdsWork.process()) {
+            yields++;
+        }
+        assertEquals(length, groupByHash.getGroupCount());
+        assertGreaterThanOrEqual(yields, (int) log(length), "at least log(length) many rehash");
+        assertEquals(length, getGroupIdsWork.getResult().getPositionCount());
+    }
+
+    @Test
+    public void testDictionaryYield()
+            throws Exception
+    {
+        // Create a page with positionCount >> expected size of groupByHash
+        int dictionaryLength = 1_000;
+        int length = 2_000;
```

Comment:
Since `length` doesn't affect the number of yields (and only dictionaryLength does). Increase length so that log(length) and log(dictionaryLength) is obviously different.